### ISSUE TITLE
Integrate Homebrew and Homebrew Cask more closely

### ISSUE THREAD

**Issue by MikeMcQuaid** [caskroom/homebrew-cask#14384]
Hi, I'm @mikemcquaid one of the Homebrew maintainers. I love Homebrew Cask and I'd love to see our projects work more closely together.

A few ideas I had:
- integrate Homebrew Cask so it's installed by Homebrew by default
- move Homebrew Cask/Caskroom repositories into the Homebrew organisation

Feel free to decline either of these. Also, feel free to suggest other things that Homebrew could do to make your life easier.

Thanks for all your great work!

**Comment by vitorgalvao** [comment]
Hello, Mike, great to have you here, as always. As you may or not already know, I [took on the role of lead maintainer a while back](https://github.com/caskroom/homebrew-cask/issues/12775), and when I did, I was told you had briefly expressed interest in this previously. The reason I mention this will become apparent in a moment, but before continuing, I’ll bring into the discussion our regular maintainers, in no particular order: @jawshooah @victorpopkov @alebcay @fanquake.

Now, on to discussing this. This will be a huge post, and I’ll bring up some things I have been thinking for a long  while about the future of homebrew-cask but haven’t discussed yet. Might as well do it now, since those pertain to a possible project merge.

First, incompatibilities between the projects. One of the things I remember reading on hombrew formulas is that you frown upon (might no longer be the case) a developer submitting a formula for their own app. We don’t, and I believe both of us are correct. I believe your decision is the right one for formulas, and that ours is the right one for casks. They are different types of applications, and should be treated differently. You also require projects to be open-source (am I recalling that correctly?) and that they are correctly tagged and versioned. We don’t, and once again I believe we’re both right. Not only because a good chunk of our apps are commercial (such is the nature of user-facing GUI apps), but because many of them simply do not offer a URL that is versioned (though I do have a possible solution in mind for that, that should at another time).

I mention all that now because you might find those differences unacceptable (which is fine), and if you do there might be no point in continuing to read this monstrosity of a post, as those aren’t really negotiable, since we need them for our users. In the end, we are different projects with different needs, and even though I’m absolutely fine with a bump in transition, regular usage should not be affected (i.e. they shouldn’t lose possibilities).

On to the projects themselves. What I deeply believe we do wrong, and should fix:

+ **Like homebrew**, we link the executables what we download (as opposed to moving or copying).
+ **Unlike homebrew**, instead of having a system that describes step-by-step *in order* how to install something, we have files that describe the package throughly in which stanza order isn’t really important.

As for the first one, [we have an agreement it needs changing](https://github.com/caskroom/homebrew-cask/issues/13201) and **for our projects to merge, this needs to be a possibility**.

As for the latter, that is the one I haven’t discussed yet but have been thinking about. As with the previously linked issue (#13201), this decision (of the way it currently works) was made based on an assumption that turned itself to be less and less true the more apps were submitted. Again, linking to [one of my favourite posts ever in the history of homebred-cask](https://github.com/caskroom/homebrew-cask/pull/259#issuecomment-17450940), being smart doesn’t work.

The reason the system works as it does now is based on the assumption that app bundles are all the same. Unlike homebrew where most install operations can be considered custom to an extent (like running `make` with certain flags, depending on user choice and the packages themselves), the reverse is true with apps: just download, unpack, link (soon to be “move”). That might very well be true for most packages, but for the ones it is not, oh what great hoops we jump through. Since our operations aren’t executed in the order they are written, [workarounds like the `*flight` stanzas](https://github.com/caskroom/homebrew-cask/blob/master/doc/CASK_LANGUAGE_REFERENCE.md#flight-stanzas-details) had to be introduced, and what a mess that sometimes turns out to be in practice. Instead of just saying in order what we want to happen, we have to define special rules that will be executed at certain times. This all gets back to #13201 and its simple idea: there are too many edge-cases, more than anyone could predict before actually starting to try to make sense of all the user-facing apps that exist for a particular OS, which is where the quote from [The Cathedral and the Bazaar](https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar) comes in again: “you often don't really understand the problem until after the first time you implement a solution. The second time, maybe you know enough to do it right. So if you want to get it right, be ready to start over at least once”.

Now, some clarifications. It’s not like homebrew does *everything* in order. Things like `url` and `homepage` (which we share) do not necessarily need to be part of the install process, no need to have to say every time “download this url, now check against this checksum”, and `homepage` wouldn’t even make sense in such a system. However, I do think that having a more descriptive installation process could do wonders for us. How far we need to go with this, I still do not know, but again, like #13201 (yes, the ideas for these two are very closely connected), this could alleviate some other burdens like [how we create shim scripts](https://github.com/caskroom/homebrew-cask/issues/9203), the aforementioned [`*flight` stanzas](https://github.com/caskroom/homebrew-cask/blob/master/doc/CASK_LANGUAGE_REFERENCE.md#flight-stanzas-details), and even the old #142, which homebrew can already do to an extent (and which would let us get rid of so many dirty parts of [caskroom/versions](https://github.com/caskroom/homebrew-versions)).

On to technical and organisational problems. The first one is, like I said before, this isn’t something I have discussed before for various reasons, so I don’t have opinions from the rest of the team, yet. What I described is the first phase of my thought process for this possible change. It would, however, **scale casks back, to some degree, to the role of formulas**. That is a big deal, and likely a huge positive. Why? Because suddenly there’s a smaller barrier to contribute to the other project if you already contribute to one of them, *and it would make homebrew-cask be able to advance faster*.

I’ll have to explain what I mean by that. Stagnation was recently a real issue for this project. You couldn’t tell from the outside because so many casks were being merged, but core features were essentially frozen. In my view, and I speak solely on my behalf here, this was due to saying “yes” to too many things, many of them I opposed for exactly this reason. This has left us with a bunch of half-baked features, and we currently still lack the people to handle it all. Truth is, we need more ruby developers to work on the core, and getting them is part of my goal. If we share more code with homebrew itself, less work we’d have there.

Speaking of that, you may be aware we at some point in the past separated from homebrew. This in the sense that we copied all the relevant code over and don’t directly depend on it. This was necessary since updates to homebrew broke homebrew-cask for everyone one too many times, and separating was the right move to keep everything working. However, it also meant we don’t get homebrew updates “for free”, and since we now have too few people working on the core, this becomes more of an issue, particularly since the migration still left some things abandoned. Again, scaling casks back to a role more akin to formulas might help us get homebrew’s code again and get the benefit of not having everything break when an update happens.

As mentioned previously, this is just the initial though of a growing idea so I don’t have all the details though out (so I haven’t even checked if everything would be possible or homebrew’s code to support this), but I was envisioning something in the real of, like you have `prefix.install_metafiles`, `bin.install`, and others, our `pkg`, `app`, `bin` (might benefit from a different name) would be ported to homebrew to do what we do. Initially I was thinking of these as specific to us, but if we follow your idea of installing homebrew-cask by default with homebrew, then might as well just make it an integral part of the system. That is, naturally, very open to debate, as is everything else.

Finally, what would happen to homebrew-cask’s team? Would we be completely autonomous under the homebrew organisation to make our own decisions regarding casks? I see this as huge since we have a bunch of experience dealing with casks and I trust the rest of the team and our ability to make the best decisions together, so it’s important to me to make sure we still have the final say in matters regarding the cask part of homebrew.

Thank you for reaching out, Mike, and thank you for all your (and the rest of homebrew’s team) work. I’m eager to know your opinions regarding this points, as well as the opinions of the other homebrew-cask maintainers.

**Comment by mikemcquaid** [comment]
> In the end, we are different projects with different needs, and even though I’m absolutely fine with a bump in transition, regular usage should not be affected (i.e. they shouldn’t lose possibilities).

I completely agree. To be clear: I think Homebrew Cask is run extremely well and I'd want you to continue to run it however you see fit. Homebrew's formula requirements are and would remain irrelevant to Cask.

> Initially I was thinking of these as specific to us, but if we follow your idea of installing homebrew-cask by default with homebrew, then might as well just make it an integral part of the system. That is, naturally, very open to debate, as is everything else.

Yup. I agree. I could see Homebrew Cask having a `CaskFormula` class (or just block, like the current way Casks are implemented) which means you don't ever need to write an `install` method (or, indeed, `test` or similar). You can use all Homebrew's internal APIs but without needing to implement formulae as we do. Essentially: all the code to actually install/interact with casks could live in Homebrew and the casks themselves could live in a tap.

> Finally, what would happen to homebrew-cask’s team? Would we be completely autonomous under the homebrew organisation to make our own decisions regarding casks? I see this as huge since we have a bunch of experience dealing with casks and I trust the rest of the team and our ability to make the best decisions together, so it’s important to me to make sure we still have the final say in matters regarding the cask part of homebrew.

You would be completely autonomous, yes. You're doing a great job and I don't want to mess with that. If the Homebrew and Cask core code merged obviously there'd be debate about the parts of the core code that overlap but by the sound of things that wouldn't be an objection. I'd propose we make a `homebrew-cask` tap which your team has complete admin rights for.

**Comment by alebcay** [comment]
> Stagnation was recently a real issue for this project. You couldn’t tell from the outside because so many casks were being merged, but core features were essentially frozen.

I've noticed this in the past few months. We've done a good job hiding it by slathering the pull requests page with a bunch of colorful tags, but the infrequent release cuts show that not much is being added to the core.

> I do think that having a more descriptive installation process could do wonders for us. How far we need to go with this, I still do not know, but again, like #13201 (yes, the ideas for these two are very closely connected), this could alleviate some other burdens like how we create shim scripts, the aforementioned *flight stanzas, and even the old #142, which homebrew can already do to an extent (and which would let us get rid of so many dirty parts of caskroom/versions).

Definitely +1 for this. There are some cases in which a procedural Cask DSL would have been very helpful. For example, I once proposed that we consider using AppleScript for getting through GUI installers/uninstallers that would be otherwise impossible, but it would be rather ugly to implement in the Cask DSL as it currently stands. If we choose to transition to a more procedural DSL, the fact that most Casks already have a fairly standardized order of stanzas would make the transition less stressful.

Another thing I like about Homebrew that I wish we had in Homebrew Cask is the special download strategies that Homebrew has. For example, the ability to specify mirrors, which have saved me when installing Formulae on several occasions, was quite nice. We've discussed the possibility of [localization stanzas](https://github.com/caskroom/homebrew-cask/issues/5592#issuecomment-50955621) for housing apps that have separate versions for different locales (e.g. Firefox, OpenOffice), but that never came to fruition, and thus, there is a pile of Casks sitting in `caskroom-versions` that needs to be updated en masse every time Mozilla pushes an update through.

If we're switching to a more procedural DSL, I would also suggest that we consider adding some sort of constructs for building apps from source code. While I think we should still prefer installing binaries over building, simply because apps are often more finicky to build than CLI software/libraries, etc., it would still be helpful for some apps that don't provide binary releases. If we have any sort of unification between the Homebrew/Homebrew Cask codebases, all that would have to be implemented to allow building apps would be some mechanism for running `xcodebuild`. Some other apps are even nice enough to provide a `Makefile` or other stuff, so that we don't even have to directly invoke `xcodebuild`. Thus, if some Homebrew methods are made available to Homebrew Cask, the additional effort to allow building apps would be minimal.

**Comment by mikemcquaid** [comment]
> Definitely +1 for this. There are some cases in which a procedural Cask DSL would have been very helpful. For example, I once proposed that we consider using AppleScript for getting through GUI installers/uninstallers that would be otherwise impossible, but it would be rather ugly to implement in the Cask DSL as it currently stands. If we choose to transition to a more procedural DSL, the fact that most Casks already have a fairly standardized order of stanzas would make the transition less stressful.

I think that if you become part of Homebrew DSL changes will require a bit more cooperation from both sides. I'd also advise against (as @vitorgalvao seems to suggest above) adding more DSL complexity for niche cases but that'd be up to the Cask team.

> If we're switching to a more procedural DSL, I would also suggest that we consider adding some sort of constructs for building apps from source code. While I think we should still prefer installing binaries over building, simply because apps are often more finicky to build than CLI software/libraries, etc., it would still be helpful for some apps that don't provide binary releases. If we have any sort of unification between the Homebrew/Homebrew Cask codebases, all that would have to be implemented to allow building apps would be some mechanism for running xcodebuild. Some other apps are even nice enough to provide a Makefile or other stuff, so that we don't even have to directly invoke xcodebuild. Thus, if some Homebrew methods are made available to Homebrew Cask, the additional effort to allow building apps would be minimal.

I think this is something I'd feel relatively strongly about: I think if Homebrew and Homebrew Cask move together then Cask should take over `homebrew-binary` and Homebrew should stick to building application bundles. I think it's a confusing distinction if both Homebrew and Homebrew Cask are building app bundles and shipping them. If Cask does ship app bundles they should be prebuilt like e.g. Homebrew bottles are and installed for end-users in the same way existing Casks are.

> The issue of whether or not to include Homebrew Cask by default seems like a trivial decision to me, because the people that need to use Cask can find it pretty easily anyways, and the percentage of people that wish they had something like Homebrew Cask but don't know about it (and are incapable of using Google to find it) is pretty slim, especially after LifeHacker did a write-up on us.

I wouldn't underestimate how this would improve your discoverability. Most random Homebrew users I've met and I've told about Homebrew Cask hadn't heard of it. I think that's a shame given how awesome a tool it is.

**Comment by vitorgalvao** [comment]
Thank you for clarifying all my questions. The more I think about it, the more I like the prospect. This would mean in a big part rewriting a lot of homebrew-cask from the ground up, which I personally see as a big win. We could implement a lot of things better if coupled to homebrew.

I’d still like to hear from @fanquake, @victorpopkov, and @jawshooah as well.

**Comment by mikemcquaid** [comment]
> Thank you for clarifying all my questions. The more I think about it, the more I like the prospect. This would mean in a big part rewriting a lot of homebrew-cask from the ground up, which I personally see as a big win. We could implement a lot of things better if coupled to homebrew.

I'd strongly suggest that you don't make that a requirement but instead we work on integrating the existing code into core first and then try and use more of Homebrew's stuff over time.

**Comment by vitorgalvao** [comment]
Yes, don’t worry, I agree we should start with what we have and then chop back what we don’t need.

**Comment by fanquake** [comment]
Sorry I haven't been keeping quite up to date with this, and haven't had a chance to read through all the comments/discussion. In general I think it's probably a good idea, as long as we maintain our autonomy, and that looks to be the case. Although, what happens as we move more & more of the cask code into Homebrew core?

If we did merge, would we have the opportunity to use, or expand and use, some of the homebrew automated testing infrastructure? I've been looking into automated updates for our versions repo, (which will also need to be renamed in a merge), and a brew bot type setup would work well for automating out some of the nightly/hourly cask updates. Although as cask becomes more complex, ideally we'll cut a lot of the casks out of versions.

**Comment by mikemcquaid** [comment]
> In general I think it's probably a good idea, as long as we maintain our autonomy, and that looks to be the case.

Definitely.

> Although, what happens as we move more & more of the cask code into Homebrew core?

There may be some loss of autonomy on the core code specifically in that case. Obviously if a Cask maintainer wants to make a core change that will cause other Homebrew issues: that won't be OK

> If we did merge, would we have the opportunity to use, or expand and use, some of the homebrew automated testing infrastructure?

Yes. That said: we're currently trying to move more of it to Travis CI due to the pain of maintaining it. That's not to say you won't be able to use it but if you use it you'll need to help maintain it :grinning:

**Comment by RandomDSdevel** [comment]
If this happens, then I'd like to see more Homebrew commands than just `brew update` automatically take care of their Homebrew Cask equivalents.  For example:  

* `brew outdated` could show outdated Casks.  
* `brew upgrade` could also run `brew cask upgrade` if and when that comes online.  
* `brew cleanup` could run `brew cask cleanup`
* `brew doctor` could run `brew cask doctor`.  

Some work might need to be done on the differences between the two systems' commands' outputs in order to make that function, though, particularly in the latter two cases.

**Comment by mikemcquaid** [comment]
@RandomDSdevel Agreed.

**Comment by RandomDSdevel** [comment]
@mikemcquaid:  Another thing that could be fixed is that the results of updating `caskroom/homebrew-cask` and `caskroom/homebrew-versions` don't show up with the results of updating the rest of your taps when you run `brew update`, and this could also be an issue for other `caskroom`-based (possibly even all `homebrew`–repository-external) formulas/Casks/etc..

**Comment by mikemcquaid** [comment]
@RandomDSdevel :+1:

**Comment by RandomDSdevel** [comment]
@mikemcquaid:  You're welcome!

**Comment by vitorgalvao** [comment]
@mikemcquaid Another question: how would we deal with our other repos? I’m concerned specifically with cases like [caskroom/versions](https://github.com/caskroom/homebrew-versions). From the way homebrew handles taps, that repo is called `homebrew-versions`, but another repo with that name already exists in homebrew. So we’d need to rename those, or there a better way of handling it?

**Comment by RandomDSdevel** [comment]
@vitorgalvao:  Didn't you say something in regard to [Issue #142](https://github.com/caskroom/homebrew-cask/issues/142) earlier on in this very same thread?  Oh, wait; here it is:  

> …However, I do think that having a more descriptive installation process could do wonders for us.  How far we need to go with this, I still do not know, but again, like #13201 (yes, the ideas for these two are very closely connected), this could alleviate some other burdens like how we create shim scripts, the aforementioned *flight stanzas, and even the old #142, which homebrew can already do to an extent (and which would let us get rid of so many dirty parts of caskroom/versions).  

But what did you mean by '`homebrew` can already do this to an extent?'

**Comment by vitorgalvao** [comment]
Why is it always the people who don’t actually contribute that keep arguing senslessly about the direction of the project like they know all its intricacies? It’s at this point we know why some developers snap and just say “fork it and build it yourself”. Though I personally won’t do it because I don’t think it is necessarily a fair answer, I unfortunately understand the sentiment.

Your comment is clearly a response to the other issue, so I’ll ask that you stop spamming other issues with it. Your sarcasm is also not appreciated. It is absolutely fine to disagree and ask for clarification. It is not OK to hijack other issues and keep screaming for the feature you want when you have been told in no unclear terms “absolutely not, but perhaps in the future, though unlikely” and have been told exactly *what* “in the future” actually means and *why* it can only be considered again at such time.

This issue/thread is about the merging of projects, not features. It is already big enough to read, and we’re still waiting on the replies of other maintainers. If you can’t take a “no” to your idea, **when the reasoning has been explained multiple times**, then I kindly ask you to leave.

**Comment by RandomDSdevel** [comment]
@vitorgalvao:  

> Why is it always the people who don’t actually contribute that keep arguing senslessly about the direction of the project like they know all its intricacies?  …

Because we want to get involved as end-users (and, hopefully, start contributing eventually,) obviously!  

> …It’s at this point we know why some developers snap and just say “fork it and build it yourself”.  Though I personally won’t do it because I don’t think it is necessarily a fair answer, I unfortunately understand the sentiment.  
>
> …

I'm sorry to have made you feel that way.  I didn't mean to, and I promise to try and be more careful in the future.  

> …
>
> Your comment is clearly a response to the other issue, …

Which other issue, [#142:  'Allow for more flexibility with multiple versions of the same app'](https://github.com/caskroom/homebrew-cask/issues/142) or [#14833:  'Application Auto-Update Interfacing and Automation](https://github.com/caskroom/homebrew-cask/issues/14833#issuecomment-153095255), the latter of which is, even though it may be what might have gotten you mad at me in the first place, irrelevant to this discussion?  I'm happy to let the latter issue sit and gather dust, and I only mentioned the _former_ issue because _you_ referenced it in the first place and I was wondering exactly why you did; I apologize for not being more clear about this.  

> …so I’ll ask that you stop spamming other issues with it.  …

I don't intend to mention it again unless somebody else believes that it has to do with this discussion first.  

> Your sarcasm is also not appreciated.  

Again, I did not mean to come off as sounding sarcastic.  

> …It is absolutely fine to disagree and ask for clarification.  

That was my original intent, and I once again apologize for not being more clear about that.  

> It is not OK to hijack other issues…

Again, I did not desire to do so.  

> …and keep screaming for the feature you want when you have been told in no unclear terms “absolutely not, but perhaps in the future, though unlikely” and have been told exactly what “in the future” actually means and why it can only be considered again at such time.  
>
> …

Please forgive for being impatient enough at the time to want a more specific answer as to what tasks needed to be dealt with prior to any reconsideration of any of my contributions, however small they are.  

> …
>
> This issue/thread is about the merging of projects, not features. …

I understand that.  I just thought that it might be advantageous for the features I've mentioned in this thread to be more consistent between the two projects even if they still maintained some differences, and @mikemcquaid apparently agrees with me on this.  

> …  It is already big enough to read, …

That's true, but I didn't think that inter-project feature consistency was that big of an addition.  If it is, then I'd be happy to move discussion of that to another thread (pending its creation if such a thread should be created by a maintainer and not an end-user like myself.)  I'd also be happy to help test said inter-project feature consistency's implementation once it starts being created.  

> …and we’re still waiting on the replies of other maintainers.  …

If that precludes end-user feedback for the time being, then I'll willingly withdraw myself from this discussion until such a time as end-user feedback is sought by said maintainers, though I still wish to follow this discussion and see how things go.  

> …  If you can’t take a “no” to your idea, **when the reasoning has been explained multiple times,** then I kindly ask you to leave.  

Please pardon me for intruding, then, as I'll leave you to work in peace.  I will, however, be available for comment if necessary.

**Comment by mikemcquaid** [comment]
> @mikemcquaid Another question: how would we deal with our other repos? I’m concerned specifically with cases like caskroom/versions. From the way homebrew handles taps, that repo is called homebrew-versions, but another repo with that name already exists in homebrew. So we’d need to rename those, or is there a better way of handling it?

@vitorgalvao We could either migrate those things to `homebrew/homebrew-versions` or have `caskroom/homebrew-versions` or whatever. Basically: any number of options.

**Comment by mikemcquaid** [comment]
> I understand that. I just thought that it might be advantageous for the features I've mentioned in this thread to be more consistent between the two projects even if they still maintained some differences, and @mikemcquaid apparently agrees with me on this.

I think this is up to the Cask team and @vitorgalvao. I figured you were part of that existing team.

> If that precludes end-user feedback for the time being, then I'll willingly withdraw myself from this discussion until such a time as end-user feedback is sought by said maintainers, though I still wish to follow this discussion and see how things go.

I think with such things end-user feedback is not really that valuable until we have technical contributions. If you'd like to contribute technically, that would be great and I'd love to help you learn.

**Comment by fanquake** [comment]
@mikemcquaid Did you have some sort of timeframe/roadmap in mind for doing this? What are going to be the major sticking points?

**Comment by mikemcquaid** [comment]
I'm prepared to do work on this and help out but the timeframe and roadmap is up to the Cask core developers.

**Comment by adityadalal924** [comment]
As a new maintainer to Homebrew Cask, I can definitely see the benefits of integrating with Homebrew. We have lots of contributors and commits, but it seems like the core of Homebrew Cask and the DSL is staying fairly static.

I'm unfortunately not very experienced with Ruby, but the idea of having a `CaskFormula` class or block so that Casks can "plug in to" Homebrew's already built up core sounds very promising, and the method  I feel would work the best for a gradual changeover.

@mikemcquaid Would you mind going into a little bit more detail about what that would entail? i.e. Would one of the Homebrew Cask core maintainers have to create this `CaskFormula` class, and then we would transition Casks to the new format, similar to a DSL change or something of the sort? (which would certainly be a lot of work up front, but at the benefit of a more stable core down the road - of course, I'm sure we can script away most of the "normal" Cask transforms)

We're still not at a "v1" release either, so it could also be an excellent opportunity to make some larger, more breaking changes (#13201 comes to mind). A change like #15218 could also be another "nice to have" sort of thing, but is something that has't come up yet.

Just some thoughts - I'm interested in getting a roadmap (albeit tentative) ready so that Homebrew Cask can continue to move forward!

**Comment by mikemcquaid** [comment]
> @mikemcquaid Would you mind going into a little bit more detail about what that would entail? i.e. Would one of the Homebrew Cask core maintainers have to create this CaskFormula class, and then we would transition Casks to the new format, similar to a DSL change or something of the sort? (which would certainly be a lot of work up front, but at the benefit of a more stable core down the road - of course, I'm sure we can script away most of the "normal" Cask transforms)

Something like that, yeh. Another version would just be to continue with the current format but just have all casks in a separate directory. We are hopefully going to separate our core and formulae at some point into separate repositories so we'd probably have another separate repository for Cask's casks and just have the core be able to read formulae or casks (and they can share code internally).

**Comment by victorpopkov** [comment]
I thought I have already posted my opinion here about integration with Homebrew more closely, but it seems not. My bad.

When I first found Homebrew Cask (HC) I was already thinking that this would be great if Homebrew and HC would be more closely integrated. So my **final answer** would be **definitely yes**.

As everyone has noticed already there weren't any breaking changes in the HC in the past few months even though the PR are always coming most of them are only targeted on new cask additions or updates. There aren't many core maintainers at the moment so nothing is added to the core itself. Integration with Homebrew will be a really good step forward to improve this situation and most likely even increase the number of contributors to the core in the future. Even from this perspective this seems to be a good idea.

I don't want to repeat the same thoughts that were posted earlier. I really like all of them, especially how @vitorgalvao has provided so many details on this subject. However, I think that first of all there is a need to have a roadmap of integration with Homebrew before making the final decision.

P.S. Thank you, @mikemcquaid, for all your precious time for answering all the questions.

**Comment by mikemcquaid** [comment]
> However, I think that first of all there is a need to have a roadmap of integration with Homebrew before making the final decision.

It's unclear to me who is :raised_hands: to do the work required here; we'd probably need volunteers before we have a roadmap.

**Comment by mikemcquaid** [comment]
So, Homebrew Cask is now a Homebrew Tap command which means Homebrew can autoupdate it. It's also installed automagically with e.g. `brew cask list` or any other `brew cask` command (thanks to @xu-cheng here).

Some next potential steps I could see us doing:
- Move this repository into the Homebrew organisation (obviously giving administrator access and priviledge to create more repos in that organisation to all the current Cask maintainers). GitHub handles automatic redirects for these nowadays.
- Try to reduce duplication between projects: https://github.com/caskroom/homebrew-cask/issues/15603
- Try and move some homebrew-cask code into Homebrew core (just as-is, for now, in a subdirectory or something) to allow Homebrew to handle casks natively and allow Cask to make use of Homebrew's core code again. @xu-cheng has also been working on separating the Homebrew core and formula code so maybe this would make sense to wait until after then. Eventually we will be able to start unifying code so that we can have semver code (perhaps even a RubyGem) that can build `Formula`e and `Cask`s on OS X and Linux with stable releases outside of `Formula`e themselves.

**Comment by fanquake** [comment]
Sounds like a good idea. Have we decided what we're going to do about our homebrew-versions repo. Can it just be renamed?

**Comment by mikemcquaid** [comment]
@fanquake I think ideally renaming it to `homebrew-cask-versions` would be good. We could always try moving e.g. it to the Homebrew org before the main `homebrew-cask` repository.

**Comment by vitorgalvao** [comment]
@mikemcquaid For some reason, I was under the impression homebrew taps couldn’t have hyphens other than the initial one (so `homebrew-cask-versions` would not work, but `homebrew-caskversions` would). Is that no longer the case, or am I just remembering it wrong and it always worked?

It’s a shame we lose the `/`. `caskroom/versions` or `cask/versions` looks way better than `cask-versions`. It’s understandable, though. It’s also important users can see at a glance (at least in the homebrew organization) if a repo will require them to type `cask` or not. In discussions like https://github.com/caskroom/homebrew-cask/issues/5876 (which I’d like to revive) if we end up with a `drivers` repo, it’d be useful for users to not confuse cask installed drivers (because they can be closed-source) with homebrew built-from-source drivers.

Would it be desirable to, in the future (perhaps after both cores are separated from the formulas and casks) to extend the “homebrew taps must be preceded with `homebrew-`” rule to “homebrew taps must be preceded with either `homebrew-` or `cask-`”? Not yet sure it would, I’m just thinking out loud, here. As I write this I see some problems with that that would need further consideration.

**Comment by mikemcquaid** [comment]
> @mikemcquaid For some reason, I was under the impression homebrew taps couldn’t have hyphens other than the initial one (so homebrew-cask-versions would not work, but homebrew-caskversions would). Is that no longer the case, or am I just remembering it wrong and it always worked?

I think it didn't used to work but it may work now (and if it still doesn't: we'll fix it so it does).

> It’s a shame we lose the /. caskroom/versions or cask/versions looks way better than cask-versions. It’s understandable, though. It’s also important users can see at a glance (at least in the homebrew organization) if a repo will require them to type cask or not. In discussions like #5876 (which I’d like to revive) if we end up with a drivers repo, it’d be useful for users to not confuse cask installed drivers (because they can be closed-source) with homebrew built-from-source drivers.

Yeh, I can see your argument here and in some ways I agree that the closed-source separate makes sense. Maybe we could split it so that we have open-source casks in `Homebrew/homebrew-cask` or something?

> Would it be desirable to, in the future (perhaps after both cores are separated from the formulas and casks) to extend the “homebrew taps must be preceded with homebrew-” rule to “homebrew taps must be preceded with either homebrew- or cask-”? Not yet sure it would, I’m just thinking out loud, here. As I write this I see some problems with that that would need further consideration.

I can see an argument for that after cores are merged.

**Comment by maxnordlund** [comment]
I've a small suggestion regarding the open-source GUI apps without official binaries. This may be an obvious suggestion, but in any case here it goes:

In the caskroom/unoffical (or Hombrew/unoffical if it's moved) repo have both a Formulae and a Cask per app. The formulae contains the knowledge to build it, and the cask how to install it on the end users machine. Essentially use the right tool for the job/rely on the strength of each.

If the brew-bot stays around for a bit longer it could build the app, but update the Cask instead.

---

On another note, I would love to help out but as an outsider I don't know where to start. Having an roadmap and maybe some documentation of the internals would really help. That being said, I don't mind learning by doing. As the say goes, if you want something done, do it yourself. (Which means now I have to write some docs :wink:)

**Comment by vitorgalvao** [comment]
@maxnordlund All help is welcome.

Your suggestion does go against the rest of the goals/functionality, though, for various reasons:

+ [We want to reduce duplication](https://github.com/caskroom/homebrew-cask/issues/15603), not add it.
+ [Mixing formulas and casks in the same repo is a big no](https://github.com/caskroom/homebrew-cask/issues/15603#issuecomment-168521947) (see replies to that comment).
+ caskroom/unofficial’s idea is very far removed from that. It’s essentially caskroom/use-at-your-own-risk, but with a shorter name. It’s the dump for correct casks that should be in the main repo but can’t for same reason ([walled downloads](https://github.com/caskroom/homebrew-cask/blob/master/CONTRIBUTING.md#finding-a-home-for-your-cask), for example).

**Comment by maxnordlund** [comment]
> * We want to reduce duplication, not add it.
Not sure if this would do that. For me it seams like you'll need to extend the Cask DSL to support building apps from source? Or is the idea to have a unified Formulae DSL?

To avoid needing to extend the Cask DSL, why not just use the existing one? That was my idea anyways, but I've may very well have missed something.

> * Mixing formulas and casks in the same repo is a big no (see replies to that comment).
Yes I did read that, and for any other repo it is definitely the right thing. But it would be done for the _unoffical_ repo. Here I would think that the Cask maintainers would still be the owners, and hopefully the Formulae would not need to be updated (perhaps just the version/sha), since they represent the recipe to build the app in question.

On the other hand, you could extend the Formulae DSL to support installing the app, if it's not already possible. But considering that brew links, and the current plan for cask is to move, it might be hard(er) to do.

> * caskroom/unofficial’s idea is very far removed from that. It’s essentially caskroom/use-at-your-own-risk, but with a shorter name. It’s the dump for correct casks that should be in the main repo but can’t for same reason (walled downloads, for example).
Well, in this thread at least it was mentioned as the place for unofficial .app bundles, or that was how I read it. If it is you're you-are-on-your-own-good-luck-have-fun repo, then it might be OK for it to have different rules regarding mixing Formulae and Cask.

I hope this explains my reasoning better, but it is just a _suggestion_ nothing more. It's hard to have a nice tone of voice in writing, as it always seams so harsh, but it's not my intention. I apologize if that were the case here.

**Comment by mikemcquaid** [comment]
> On the other hand, you could extend the Formulae DSL to support installing the app, if it's not already possible. But considering that brew links, and the current plan for cask is to move, it might be hard(er) to do.

It is already possible; they just aren't linked automatically and require running `brew linkapps`. If we have  a dedicated tap for these things I think we could also run `linkapps` automatically too.

**Comment by vitorgalvao** [comment]
> For me it seams like you'll need to extend the Cask DSL to support building apps from source?

That’s the last thing we need to do. It’s *way* out of our goals, and it’ll likely never happen. There’s no point to it, considering what we want for this tool. In addition, homebrew can already do all that (and they do for a few apps).

Regarding, caskroom/unofficial:

> If it is you're you-are-on-your-own-good-luck-have-fun repo

That’s not what I said. I said “use-at-your-own-risk”. Difference being your wording suggests we’d offer no help/support with it, but that’s not the case. What I meant is though we do our best to have everything correct, there’s something that prevents us from doing so in those cases. We can’t verify the authenticity of their download urls, for example, so they might pose a security risk and you should proceed with caution.

> then it might be OK for it to have different rules regarding mixing Formulae and Cask.

I stand by my comments in the other issue: different things, different structures, different goals, maintenance nightmare with no visible gains. Not worth it in the least.

> I hope this explains my reasoning better, but it is just a suggestion nothing more. It's hard to have a nice tone of voice in writing, as it always seams so harsh, but it's not my intention. I apologize if that were the case here.

It was not, but I know the feeling, so now I’ll have to apologise from my side. Many times I also sound harsh in writing, but that is almost never my intention.

**Comment by maxnordlund** [comment]
> > On the other hand, you could extend the Formulae DSL to support installing the app, if it's not already possible. But considering that brew links, and the current plan for cask is to move, it might be hard(er) to do.

> It is already possible; they just aren't linked automatically and require running brew linkapps. If we have a dedicated tap for these things I think we could also run linkapps automatically too.

I know this may a fair bit to late but I got reminded by #14053 and I just wanted to say that I think @mikemcquaid nailed it. Use the existing linkapps, and just special case a tap.

**Comment by alebcay** [comment]
@maxnordlund The general consensus has been reached that linking simply doesn't cut it for some Casks, and it's gotten to a point that such Casks are no longer corner cases that we can conveniently ignore. There are enough Casks that break in some way when being linked that [we're eventually switching to copying them instead](https://github.com/caskroom/homebrew-cask/issues/13201).

`linkapps` is neat, but linking simply isn't dependable enough for a project such as Homebrew Cask, whose sole project is to provide apps that, often (as we have slowly discovered), hate being linked. On the other hand, it makes sense for Homebrew to have it in order to link the occasional app that is included.

**Comment by mikemcquaid** [comment]
> linkapps is neat, but linking simply isn't dependable enough for a project such as Homebrew Cask, whose sole project is to provide apps that, often (as we have slowly discovered), hate being linked. On the other hand, it makes sense for Homebrew to have it in order to link the occasional app that is included.

The difference between the two is actually somewhat more complex: all of Cask's .apps are self-contained so can me linked/copied/moved anywhere (although they generally are all intended to be put in `/Applications`). All (or almost all) of Homebrew's built-from-source .apps are not self-contained so cannot be copied or moved elsewhere so need to be linked.

**Comment by maxnordlund** [comment]
I would also add that it should be possible using configure flags/patches to make them link safe.

**Comment by fuzzy76** [comment]
Cask deals primarily with binary files, not building from source.

**Comment by mikemcquaid** [comment]
> I would also add that it should be possible using configure flags/patches to make them link safe, since we're building them from source.

Without going into too much detail: it's a lot more complex to that; it's not about being link safe but having the bundle be relocatable and self-contained.

**Comment by toonetown** [comment]
Just as a question, would this (eventually) change the need to use `cask` on the command line?  i.e. would you be able to `brew install some-cask`, or would you envision always running `brew cask install some-cask`?

I don't have a preference or an opinion one way or the other...it was just a thought that popped into my mind, and I was wondering what the plans would be around that.

**Comment by toonetown** [comment]
Never mind my question - I just saw [this comment](https://github.com/caskroom/homebrew-cask/issues/15603#issuecomment-162535599) on another thread which basically answers it.

**Comment by MikeMcQuaid** [comment]
This was done ✨
